File: lfu.py
# time-O(1)
# space-O(k)-hashmaps
# two hashmaps are maintained- one to keep the frequency- freq as key and val in the form of DLL(for the order) and the next is a mapping of key to node of(key,val,freq)
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.next = None
        self.prev = None
        self.count = 1


class DLL:
    def __init__(self):
        self.size = 0
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def add(self, node):
        node.next = self.head.next
        self.head.next = node
        node.next.prev = node
        node.prev = self.head
        self.size += 1

    def remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
        self.size -= 1

    def removelast(self):
        last = self.tail.prev
        self.remove(last)
        return last


class LFUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.map = {}
        self.freqmap = {}
        self.min = float('inf')

    def update(self, node):
        count = node.count
        dll = self.freqmap[count]
        dll.remove(node)
        if dll.size == 0 and self.min == count:
            self.min += 1
        count += 1
        if count not in self.freqmap:
            newlist = DLL()
            self.freqmap[count] = newlist
        newdll = self.freqmap[count]
        newdll.add(node)

    def get(self, key: int) -> int:
        if key in self.map:
            node = self.map[key]
            self.update(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return
        if key in self.map:
            node = self.map[key]
            node.val = value
            self.update(node)
        else:
            node = Node(key, value)
            if len(self.map) == self.capacity:
                minnodes = self.freqmap[self.min]
                last = minnodes.removelast()
                del self.map[last.key]
            print(key, value)

            if 1 not in self.freqmap:
                newlist = DLL()
                self.freqmap[1] = newlist
            newlist = self.freqmap[1]
            newlist.add(node)
            self.map[key] = node
            self.min = 1
# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: src/LFU.java
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
//Time Complexity : O(1)
//Space Complexity : O(capacity)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : No

/**
 * Maintain two maps. One for the key to (freq, value) and one for frequencies
 * (freq -> set of keys). So when get is called check if there is a key in the
 * cache. If so, get its frequency and fetch the set of keys having that
 * frequency. Remove this key from that list and if that list becomes empty
 * remove that frequency from the frequencies map. Then update the key in the
 * cache with frequency+1 and value and also put the frequency with the key in
 * its set.
 * 
 * When put is called, check if there exists a key already in cache. If so,
 * update the key with the new value and old frequency and then call get of this
 * key. It will update the frequency accordingly. If the capacity is already
 * full, then get the min frequency from the frequencies map and remove it from
 * cache as well. Then insert the new key with frequency as 1. Update min to 1.
 *
 */
class LFUCache {
	Map<Integer, Pair<Integer, Integer>> cache;
	Map<Integer, LinkedHashSet<Integer>> frequencies;
	int min;
	int capacity;

	public LFUCache(int capacity) {
		this.capacity = capacity;
		this.cache = new HashMap<>();
		frequencies = new HashMap<>();
		min = 0;
	}

	public int get(int key) {
		Pair<Integer, Integer> freq = cache.get(key);
		if (freq == null) {
			return -1;
		}
		int frequency = freq.getKey();
		Set<Integer> keys = frequencies.get(frequency);
		keys.remove(key);
		if (keys.isEmpty()) {
			frequencies.remove(frequency);
			if (min == frequency)
				min++;
		}
		int value = freq.getValue();
		insert(key, frequency + 1, value);
		return value;
	}

	public void insert(int key, int frequency, int value) {
		cache.put(key, new Pair<>(frequency, value));
		frequencies.putIfAbsent(frequency, new LinkedHashSet<>());
		frequencies.get(frequency).add(key);
	}

	public void put(int key, int value) {
		if (capacity <= 0) {
			return;
		}
		Pair<Integer, Integer> freq = cache.get(key);
		if (freq != null) {
			cache.put(key, new Pair<>(freq.getKey(), value));
			get(key);
			return;
		}
		if (capacity == cache.size()) {
			Set<Integer> keys = frequencies.get(min);
			int keyToDel = keys.iterator().next();
			cache.remove(keyToDel);
			keys.remove(keyToDel);
			if (keys.isEmpty()) {
				frequencies.remove(min);
			}
		}
		min = 1;
		insert(key, 1, value);
	}
}

/**
 * Your LFUCache object will be instantiated and called as such: LFUCache obj =
 * new LFUCache(capacity); int param_1 = obj.get(key); obj.put(key,value);
 */
